contracts/Pool.sol

```diff
@@ -13,255 +15,368 @@ import "./IBorrowersRegistry.sol";
-    }
+    _updateRates();
 
+    emit Repayment(msg.sender, msg.value, block.timestamp);
+  }
 
-    /**
-     * Returns the current interest rate for borrowings
-    **/
-    function getBorrowingRate() public view returns(uint256) {
-      return ratesCalculator.calculateBorrowingRate(totalBorrowed, totalDeposited);
-    }
 
+  /* =========
 
-    /* ========== INTERNAL FUNCTIONS ========== */
 
+  /**
+   * Returns the current borrowed amount for the given user
+   * The value includes the interest rates owned at the current moment
+   * @dev _user the address of queried borrower
+  **/
+  function getBorrowed(address _user) public view returns (uint256) {
+    return borrowIndex.getIndexedValue(borrowed[_user], _user);
+  }
 
-    function updateRates() internal {
-      depositIndex.setRate(ratesCalculator.calculateDepositRate(totalBorrowed, totalDeposited));
-      borrowIndex.setRate(ratesCalculator.calculateBorrowingRate(totalBorrowed, totalDeposited));
-    }
+  function totalSupply() public view override returns (uint256) {
+    return balanceOf(address(this));
+  }
 
+  function totalBorrowed() public view returns (uint256) {
+    return getBorrowed(address(this));
+  }
 
-    function accumulateDepositInterests(address user) internal {
-        uint256 depositedWithInterests = getDeposits(user);
-        uint256 interests = depositedWithInterests - deposits[user];
-        deposits[user] = depositedWithInterests;
-        totalDeposited = totalDeposited + interests;
-        depositIndex.updateUser(user);
-    }
 
+  /**
+   * Returns the current deposited amount for the given user
+   * The value includes the interest rates earned at the current moment
+   * @dev _user the address of queried depositor
+  **/
+  function balanceOf(address user) public view override returns (uint256) {
+    return depositIndex.getIndexedValue(_deposited[user], user);
+  }
 
-    function accumulateBorrowingInterests(address user) internal {
-        uint256 borrowedWithInterests = getBorrowed(user);
-        uint256 interests = borrowedWithInterests - borrowed[user];
-        borrowed[user] = borrowedWithInterests;
-        totalBorrowed = totalBorrowed + interests;
-        borrowIndex.updateUser(user);
-    }
 
+  /**
+   * Returns the current interest rate for deposits
+  **/
+  function getDepositRate() public view returns (uint256) {
+    return _ratesCalculator.calculateDepositRate(totalBorrowed(), totalSupply());
+  }
+
+
+  /**
+   * Returns the current interest rate for borrowings
+  **/
+  function getBorrowingRate() public view returns (uint256) {
+    return _ratesCalculator.calculateBorrowingRate(totalBorrowed(), totalSupply());
+  }
+
+
+  /**
+   * Recovers the residual funds resultant from difference between deposit and borrowing rates
+  **/
+  function recover(uint256 amount, address account) public onlyOwner nonReentrant {
+    uint256 residual = address(this).balance - totalSupply();
```

Comment:
Not sure about that

Reply:
maybe something like that:
```
    require(amount <= address(this).balance, "Trying to recover more residual funds than current pool balance");

    uint256 residual = totalBorrowed() - totalSupply();
    require(amount <= residual, "Trying to recover more residual funds than available");
    ```

Reply:
Your proposed change look good to me. The actual residual seems to be the `totalBorrowed() - totalSupply();` and additionally we're checking if "current pool utilisation" (amount <= address(this).balance)  allows us to `recover`  the specified `amount` at a given time.

Reply:
Maybe let's call it recoverSurplus...